src/sbvr-api/translations.ts

```diff
@@ -0,0 +1,218 @@
+	SelectQueryNode,
+	NumberTypeNodes,
+	BooleanTypeNodes,
+	UnknownTypeNodes,
+	NullNode,
+} from '@balena/abstract-sql-compiler';
+import { Dictionary } from './common-types';
+
+export type AliasValidNodeType =
+	| ReferencedFieldNode
+	| SelectQueryNode
+	| NumberTypeNodes
+	| BooleanTypeNodes
+	| UnknownTypeNodes
+	| NullNode;
+const aliasFields = (
+	abstractSqlModel: AbstractSqlModel,
+	resourceName: string,
+	aliases: Dictionary<string | AliasValidNodeType>,
+): SelectNode[1] => {
+	const fieldNames = abstractSqlModel.tables[resourceName].fields.map(
+		({ fieldName }) => fieldName,
+	);
+	const nonexistentFields = _.difference(Object.keys(aliases), fieldNames, [
+		'$toResource',
+	]);
+	if (nonexistentFields.length > 0) {
+		throw new Error(
+			`Tried to alias non-existent fields: '${nonexistentFields.join(', ')}'`,
+		);
+	}
+	return fieldNames.map(
+		(fieldName): AliasNode<AliasValidNodeType> | ReferencedFieldNode => {
+			const alias = aliases[fieldName];
+			if (alias) {
+				if (typeof alias === 'string') {
+					return ['Alias', ['ReferencedField', resourceName, alias], fieldName];
+				}
+				return ['Alias', alias, fieldName];
+			}
+			return ['ReferencedField', resourceName, fieldName];
+		},
+	);
+};
+
+const aliasResource = (
+	abstractSqlModel: AbstractSqlModel,
+	resourceName: string,
+	toResource: string,
+	aliases: Dictionary<string | AliasValidNodeType>,
+): Definition => {
+	if (!abstractSqlModel.tables[toResource]) {
+		throw new Error(`Tried to alias to a non-existent resource: ${toResource}`);
+	}
+	return {
+		abstractSql: [
+			'SelectQuery',
+			['Select', aliasFields(abstractSqlModel, resourceName, aliases)],
+			['From', ['Alias', ['Resource', toResource], resourceName]],
+		],
+	};
+};
+
+const namespaceRelationships = (
+	relationships: Relationship,
+	alias: string,
+): void => {
+	for (const [key, relationship] of Object.entries(
+		relationships as RelationshipInternalNode,
+	)) {
+		if (key === '$') {
+			return;
+		}
+
+		let mapping = (relationship as RelationshipLeafNode).$;
+		if (mapping != null && mapping.length === 2) {
+			if (!key.includes('$')) {
+				mapping = _.cloneDeep(mapping);
+				mapping[1]![0] = `${mapping[1]![0]}$${alias}`;
+				(relationships as RelationshipInternalNode)[`${key}$${alias}`] = {
+					$: mapping,
+				};
+				delete (relationships as RelationshipInternalNode)[key];
+			}
+		}
+		namespaceRelationships(relationship, alias);
+	}
+};
+
+export const translateAbstractSqlModel = (
+	fromAbstractSqlModel: AbstractSqlModel,
+	toAbstractSqlModel: AbstractSqlModel,
+	fromVersion: string,
+	toVersion: string,
+	definitions: Dictionary<
+		| (Definition & { $toResource?: string })
+		| Dictionary<string | AliasValidNodeType>
+	> = {},
+): Dictionary<string> => {
+	const isDefinition = (d: (typeof definitions)[string]): d is Definition =>
+		'abstractSql' in d;
+
+	const resourceRenames: Dictionary<string> = {};
+
+	fromAbstractSqlModel.rules = toAbstractSqlModel.rules;
```

Comment:
Just for my understanding, the latest model rules are applied to all older models?

Reply:
No, `validateModel` will fail if called on a "virtual" model and the rules are only actually run for the latest/physical model. This is more for completeness sake of copying stuff across and might be necessary for resolving check constraint names to the relevant rule (I'm not sure which model that resolution uses)